File: audio analyzer/audio_analyzer.py
#!/usr/bin/env python3
"""
Audio Transcription and Analysis Console Application

This application transcribes audio files using OpenAI's Whisper API,
summarizes the content using GPT, and provides detailed analytics.
"""

import os
import json
import argparse
import sys
import time
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Tuple, Any
import re

# Third-party imports
try:
    import openai
    from openai import OpenAI, RateLimitError, APIError, APIConnectionError
except ImportError:
    print("Error: OpenAI library not installed. Please run: pip install openai")
    sys.exit(1)

try:
    from mutagen import File as MutagenFile
except ImportError:
    print("Error: Mutagen library not installed. Please run: pip install mutagen")
    sys.exit(1)

try:
    from dotenv import load_dotenv
    # Load environment variables from .env file if it exists
    load_dotenv()
except ImportError:
    print("Warning: python-dotenv not installed. Using system environment variables only.")
    print("For easier setup, install with: pip install python-dotenv")


# Constants for audio file validation
SUPPORTED_AUDIO_FORMATS = {'.mp3', '.mp4', '.mpeg', '.mpga', '.m4a', '.wav', '.webm', '.flac'}
MAX_FILE_SIZE_MB = 25
MAX_FILE_SIZE_BYTES = MAX_FILE_SIZE_MB * 1024 * 1024


class AudioAnalyzer:
    """Main class for audio transcription and analysis."""
    
    def __init__(self, api_key: str):
        """Initialize the AudioAnalyzer with OpenAI API key."""
        self.client = OpenAI(api_key=api_key)
        self.output_dir = Path("output")
        self.output_dir.mkdir(exist_ok=True)
    
    def validate_audio_file(self, audio_path: str) -> None:
        """Validate audio file format and size."""
        audio_file = Path(audio_path)
        
        # Check if file exists
        if not audio_file.exists():
            raise FileNotFoundError(f"Audio file not found: {audio_path}")
        
        # Check file format
        file_extension = audio_file.suffix.lower()
        if file_extension not in SUPPORTED_AUDIO_FORMATS:
            raise ValueError(
                f"Unsupported audio format: {file_extension}\n"
                f"Supported formats: {', '.join(sorted(SUPPORTED_AUDIO_FORMATS))}"
            )
        
        # Check file size
        file_size = audio_file.stat().st_size
        file_size_mb = file_size / (1024 * 1024)
        if file_size > MAX_FILE_SIZE_BYTES:
            raise ValueError(
                f"File too large: {file_size_mb:.1f}MB\n"
                f"Maximum allowed size: {MAX_FILE_SIZE_MB}MB"
            )
        
        print(f"✓ Audio file validation passed:")
        print(f"  - Format: {file_extension}")
        print(f"  - Size: {file_size_mb:.1f}MB")
    
    def get_audio_duration(self, audio_path: str) -> float:
        """Get audio duration in seconds using mutagen."""
        try:
            audio_file = MutagenFile(audio_path)
            if audio_file is not None and hasattr(audio_file, 'info'):
                return float(audio_file.info.length)
            else:
                print(f"Warning: Could not determine audio duration for {audio_path}")
                return 0.0
        except Exception as e:
            print(f"Warning: Error getting audio duration: {e}")
            return 0.0
    
    def transcribe_audio(self, audio_path: str) -> str:
        """Transcribe audio using OpenAI Whisper API with retry logic."""
        print(f"Transcribing audio file: {audio_path}")
        
        max_retries = 3
        base_delay = 1  # Base delay in seconds
        
        for attempt in range(max_retries):
            try:
                with open(audio_path, "rb") as audio_file:
                    transcript = self.client.audio.transcriptions.create(
                        model="whisper-1",
                        file=audio_file,
                        response_format="text"
                    )
                print("✓ Transcription completed successfully")
                return transcript
                
            except RateLimitError as e:
                if attempt == max_retries - 1:
                    raise Exception(f"Rate limit exceeded after {max_retries} attempts: {e}")
                
                delay = base_delay * (2 ** attempt)  # Exponential backoff
                print(f"⚠️  Rate limit hit. Retrying in {delay} seconds... (attempt {attempt + 1}/{max_retries})")
                time.sleep(delay)
                
            except (APIError, APIConnectionError) as e:
                if attempt == max_retries - 1:
                    raise Exception(f"API error after {max_retries} attempts: {e}")
                
                delay = base_delay * (2 ** attempt)
                print(f"⚠️  API error. Retrying in {delay} seconds... (attempt {attempt + 1}/{max_retries})")
                time.sleep(delay)
                
            except Exception as e:
                print(f"Error during transcription: {e}")
                raise
    
    def summarize_text(self, text: str) -> str:
        """Summarize text using OpenAI GPT model."""
        print("Generating summary...")
        
        try:
            response = self.client.chat.completions.create(
                model="gpt-4",
                messages=[
                    {
                        "role": "system",
                        "content": "You are a helpful assistant that creates concise, informative summaries of transcribed audio content. Focus on key points, main topics, and important details."
                    },
                    {
                        "role": "user",
                        "content": f"Please provide a comprehensive summary of the following transcribed audio content:\n\n{text}"
                    }
                ],
                max_tokens=1000,
                temperature=0.3
            )
            summary = response.choices[0].message.content
            print("✓ Summary generated successfully")
            return summary
        except Exception as e:
            print(f"Error during summarization: {e}")
            raise
    
    def extract_topics(self, text: str) -> List[Dict[str, Any]]:
        """Extract frequently mentioned topics using GPT."""
        print("Extracting topics...")
        
        try:
            response = self.client.chat.completions.create(
                model="gpt-4",
                messages=[
                    {
                        "role": "system",
                        "content": """You are an expert at analyzing text and identifying key topics and themes. 
                        Return a JSON array of the most frequently mentioned topics with their mention counts.
                        Each item should have 'topic' and 'mentions' fields.
                        Focus on meaningful topics, not common words like 'the', 'and', etc.
                        Return at least 3 topics, more if there are significant ones."""
                    },
                    {
                        "role": "user", 
                        "content": f"Analyze this transcribed audio content and identify the most frequently mentioned topics with their approximate mention counts:\n\n{text}\n\nReturn only a JSON array in this format: " + '[{"topic": "Topic Name", "mentions": 5}]'
                    }
                ],
                max_tokens=500,
                temperature=0.1
            )
            
            topics_text = response.choices[0].message.content
            # Extract JSON from the response
            json_match = re.search(r'\[.*\]', topics_text, re.DOTALL)
            if json_match:
                topics = json.loads(json_match.group())
            else:
                # Fallback: try to parse the entire response as JSON
                topics = json.loads(topics_text)
            
            print("✓ Topics extracted successfully")
            return topics
        except Exception as e:
            print(f"Warning: Error extracting topics with GPT: {e}")
            # Fallback to simple keyword extraction
            return self._simple_topic_extraction(text)
    
    def _simple_topic_extraction(self, text: str) -> List[Dict[str, Any]]:
        """Fallback simple topic extraction based on word frequency."""
        print("Using fallback topic extraction...")
        
        # Clean and split text
        words = re.findall(r'\b[a-zA-Z]{3,}\b', text.lower())
        
        # Common stop words to filter out
        stop_words = {
            'the', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for', 'of', 'with',
            'by', 'from', 'as', 'is', 'was', 'are', 'were', 'be', 'been', 'have',
            'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could', 'should',
            'may', 'might', 'must', 'can', 'this', 'that', 'these', 'those', 'i',
            'you', 'he', 'she', 'it', 'we', 'they', 'me', 'him', 'her', 'us', 'them',
            'my', 'your', 'his', 'her', 'its', 'our', 'their', 'a', 'an', 'some',
            'any', 'all', 'each', 'every', 'no', 'not', 'very', 'so', 'just', 'now',
            'then', 'than', 'only', 'also', 'back', 'other', 'many', 'much', 'more',
            'most', 'good', 'new', 'first', 'last', 'long', 'great', 'little', 'own',
            'same', 'right', 'big', 'high', 'different', 'small', 'large', 'next',
            'early', 'young', 'important', 'few', 'public', 'bad', 'same', 'able'
        }
        
        # Count word frequencies
        word_counts = {}
        for word in words:
            if word not in stop_words and len(word) > 3:
                word_counts[word] = word_counts.get(word, 0) + 1
        
        # Get top topics
        sorted_words = sorted(word_counts.items(), key=lambda x: x[1], reverse=True)
        topics = [{"topic": word.title(), "mentions": count} for word, count in sorted_words[:5]]
        
        return topics
    
    def calculate_analytics(self, text: str, audio_duration: float) -> Dict[str, Any]:
        """Calculate comprehensive analytics for the transcribed text."""
        print("Calculating analytics...")
        
        # Word count
        words = text.split()
        word_count = len(words)
        
        # Speaking speed (WPM)
        if audio_duration > 0:
            duration_minutes = audio_duration / 60
            speaking_speed = round(word_count / duration_minutes, 1)
        else:
            speaking_speed = 0
        
        # Extract topics
        topics = self.extract_topics(text)
        
        analytics = {
            "word_count": word_count,
            "speaking_speed_wpm": speaking_speed,
            "audio_duration_seconds": round(audio_duration, 2),
            "frequently_mentioned_topics": topics
        }
        
        print("✓ Analytics calculated successfully")
        return analytics
    
    def save_files(self, transcription: str, summary: str, analytics: Dict[str, Any], 
                   original_filename: str) -> Tuple[str, str, str]:
        """Save transcription, summary, and analytics to separate files."""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        base_name = Path(original_filename).stem
        
        # File paths
        transcription_path = self.output_dir / f"{base_name}_{timestamp}_transcription.md"
        summary_path = self.output_dir / f"{base_name}_{timestamp}_summary.md"
        analytics_path = self.output_dir / f"{base_name}_{timestamp}_analysis.json"
        
        # Save transcription
        with open(transcription_path, 'w', encoding='utf-8') as f:
            f.write(f"# Transcription\n\n")
            f.write(f"**File:** {original_filename}\n")
            f.write(f"**Date:** {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n\n")
            f.write(f"## Content\n\n{transcription}\n")
        
        # Save summary
        with open(summary_path, 'w', encoding='utf-8') as f:
            f.write(f"# Summary\n\n")
            f.write(f"**File:** {original_filename}\n")
            f.write(f"**Date:** {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n\n")
            f.write(f"## Summary\n\n{summary}\n")
        
        # Save analytics
        with open(analytics_path, 'w', encoding='utf-8') as f:
            json.dump(analytics, f, indent=2, ensure_ascii=False)
        
        print(f"✓ Files saved:")
        print(f"  - Transcription: {transcription_path}")
        print(f"  - Summary: {summary_path}")
        print(f"  - Analytics: {analytics_path}")
        
        return str(transcription_path), str(summary_path), str(analytics_path)
    
    def process_audio(self, audio_path: str) -> None:
        """Main method to process audio file completely."""
        print(f"\n{'='*60}")
        print(f"AUDIO TRANSCRIPTION AND ANALYSIS")
        print(f"{'='*60}")
        print(f"Processing: {audio_path}")
        print(f"{'='*60}\n")
        
        try:
            # Validate audio file (format, size, existence)
            self.validate_audio_file(audio_path)
            
            # Get audio duration
            audio_duration = self.get_audio_duration(audio_path)
            print(f"Audio duration: {audio_duration:.2f} seconds\n")
            
            # Step 1: Transcribe audio
            transcription = self.transcribe_audio(audio_path)
            
            # Step 2: Generate summary
            summary = self.summarize_text(transcription)
            
            # Step 3: Calculate analytics
            analytics = self.calculate_analytics(transcription, audio_duration)
            
            # Step 4: Save files
            trans_path, summ_path, anal_path = self.save_files(
                transcription, summary, analytics, os.path.basename(audio_path)
            )
            
            # Step 5: Display results
            self.display_results(transcription, summary, analytics)
            
        except Exception as e:
            print(f"❌ Error processing audio: {e}")
            sys.exit(1)
    
    def display_results(self, transcription: str, summary: str, analytics: Dict[str, Any]) -> None:
        """Display results in the console."""
        print(f"\n{'='*60}")
        print("RESULTS")
        print(f"{'='*60}")
        
        print(f"\n📊 ANALYTICS:")
        print(f"{'='*30}")
        print(json.dumps(analytics, indent=2, ensure_ascii=False))
        
        print(f"\n📝 SUMMARY:")
        print(f"{'='*30}")
        print(summary)
        
        print(f"\n📄 TRANSCRIPTION (first 500 chars):")
        print(f"{'='*30}")
        print(transcription[:500] + ("..." if len(transcription) > 500 else ""))
        
        print(f"\n{'='*60}")
        print("✅ Processing completed successfully!")
        print(f"{'='*60}")


def main():
    """Main function to handle command line arguments and run the application."""
    parser = argparse.ArgumentParser(
        description="Audio Transcription and Analysis Tool",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python audio_analyzer.py audio.mp3
  python audio_analyzer.py /path/to/audio.wav
  python audio_analyzer.py recording.m4a --api-key YOUR_API_KEY
        """
    )
    
    parser.add_argument(
        "audio_file",
        help="Path to the audio file to transcribe and analyze"
    )
    
    parser.add_argument(
        "--api-key",
        help="OpenAI API key (can also be set via OPENAI_API_KEY environment variable)"
    )
    
    args = parser.parse_args()
    
    # Get API key
    api_key = args.api_key or os.getenv("OPENAI_API_KEY")
    if not api_key:
        print("❌ Error: OpenAI API key required!")
        print("Set it via:")
        print("  - Environment variable: export OPENAI_API_KEY='your-key-here'")
        print("  - Command line argument: --api-key YOUR_KEY")
        sys.exit(1)
    
    # Validate audio file
    if not os.path.exists(args.audio_file):
        print(f"❌ Error: Audio file not found: {args.audio_file}")
        sys.exit(1)
    
    # Process audio
    analyzer = AudioAnalyzer(api_key)
    analyzer.process_audio(args.audio_file)


if __name__ == "__main__":
    main()

File: audio_analyzer/audio_analyzer.js
#!/usr/bin/env node

/**
 * Audio Transcription and Analysis Console Application
 * 
 * This application transcribes audio files using OpenAI's Whisper API,
 * summarizes the content using GPT, and provides detailed analytics.
 */

const fs = require('fs').promises;
const path = require('path');
const { program } = require('commander');
require('dotenv').config();

// Third-party imports with error handling
let OpenAI, musicMetadata;
try {
    OpenAI = require('openai');
} catch (error) {
    console.error('Error: OpenAI library not installed. Please run: npm install openai');
    process.exit(1);
}

try {
    musicMetadata = require('music-metadata');
} catch (error) {
    console.error('Error: music-metadata library not installed. Please run: npm install music-metadata');
    process.exit(1);
}

// Constants for audio file validation
const SUPPORTED_AUDIO_FORMATS = new Set(['.mp3', '.mp4', '.mpeg', '.mpga', '.m4a', '.wav', '.webm', '.flac']);
const MAX_FILE_SIZE_MB = 25;
const MAX_FILE_SIZE_BYTES = MAX_FILE_SIZE_MB * 1024 * 1024;

/**
 * Main class for audio transcription and analysis
 */
class AudioAnalyzer {
    constructor(apiKey) {
        this.client = new OpenAI({ apiKey });
        this.outputDir = path.join(process.cwd(), 'output');
        this.ensureOutputDir();
    }

    /**
     * Ensure output directory exists
     */
    async ensureOutputDir() {
        try {
            await fs.mkdir(this.outputDir, { recursive: true });
        } catch (error) {
            if (error.code !== 'EEXIST') {
                throw error;
            }
        }
    }

    /**
     * Validate audio file format and size
     */
    async validateAudioFile(audioPath) {
        // Check if file exists
        try {
            await fs.access(audioPath);
        } catch (error) {
            throw new Error(`Audio file not found: ${audioPath}`);
        }

        const audioFile = path.parse(audioPath);
        const fileExtension = audioFile.ext.toLowerCase();

        // Check file format
        if (!SUPPORTED_AUDIO_FORMATS.has(fileExtension)) {
            const supportedFormats = Array.from(SUPPORTED_AUDIO_FORMATS).sort().join(', ');
            throw new Error(
                `Unsupported audio format: ${fileExtension}\n` +
                `Supported formats: ${supportedFormats}`
            );
        }

        // Check file size
        const stats = await fs.stat(audioPath);
        const fileSizeMB = stats.size / (1024 * 1024);
        
        if (stats.size > MAX_FILE_SIZE_BYTES) {
            throw new Error(
                `File too large: ${fileSizeMB.toFixed(1)}MB\n` +
                `Maximum allowed size: ${MAX_FILE_SIZE_MB}MB`
            );
        }

        console.log('✓ Audio file validation passed:');
        console.log(`  - Format: ${fileExtension}`);
        console.log(`  - Size: ${fileSizeMB.toFixed(1)}MB`);
    }

    /**
     * Get audio duration in seconds using music-metadata
     */
    async getAudioDuration(audioPath) {
        try {
            const metadata = await musicMetadata.parseFile(audioPath);
            const duration = metadata.format.duration || 0;
            return duration;
        } catch (error) {
            console.warn(`Warning: Could not determine audio duration: ${error.message}`);
            return 0.0;
        }
    }

    /**
     * Transcribe audio using OpenAI Whisper API with retry logic
     */
    async transcribeAudio(audioPath) {
        console.log(`Transcribing audio file: ${audioPath}`);
        
        const maxRetries = 3;
        const baseDelay = 1000; // Base delay in milliseconds
        
        for (let attempt = 0; attempt < maxRetries; attempt++) {
            try {
                const audioBuffer = await fs.readFile(audioPath);
                const audioFile = new File([audioBuffer], path.basename(audioPath), {
                    type: this.getMimeType(path.extname(audioPath))
                });

                const transcript = await this.client.audio.transcriptions.create({
                    model: 'whisper-1',
                    file: audioFile,
                    response_format: 'text'
                });

                console.log('✓ Transcription completed successfully');
                return transcript;
                
            } catch (error) {
                if (this.isRetryableError(error) && attempt < maxRetries - 1) {
                    const delay = baseDelay * Math.pow(2, attempt);
                    console.warn(`⚠️  ${error.message}. Retrying in ${delay/1000} seconds... (attempt ${attempt + 1}/${maxRetries})`);
                    await this.sleep(delay);
                } else {
                    throw new Error(`Transcription failed after ${maxRetries} attempts: ${error.message}`);
                }
            }
        }
    }

    /**
     * Check if error is retryable
     */
    isRetryableError(error) {
        const retryableErrors = ['rate_limit_exceeded', 'api_error', 'connection_error', 'timeout'];
        return retryableErrors.some(type => 
            error.message.toLowerCase().includes(type) || 
            error.code === 'rate_limit_exceeded' ||
            error.status >= 500
        );
    }

    /**
     * Get MIME type for audio file extension
     */
    getMimeType(extension) {
        const mimeTypes = {
            '.mp3': 'audio/mpeg',
            '.mp4': 'audio/mp4',
            '.mpeg': 'audio/mpeg',
            '.mpga': 'audio/mpeg',
            '.m4a': 'audio/mp4',
            '.wav': 'audio/wav',
            '.webm': 'audio/webm',
            '.flac': 'audio/flac'
        };
        return mimeTypes[extension.toLowerCase()] || 'audio/mpeg';
    }

    /**
     * Sleep for specified milliseconds
     */
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    /**
     * Summarize text using OpenAI GPT model
     */
    async summarizeText(text) {
        console.log('Generating summary...');
        
        try {
            const response = await this.client.chat.completions.create({
                model: 'gpt-4',
                messages: [
                    {
                        role: 'system',
                        content: 'You are a helpful assistant that creates concise, informative summaries of transcribed audio content. Focus on key points, main topics, and important details.'
                    },
                    {
                        role: 'user',
                        content: `Please provide a comprehensive summary of the following transcribed audio content:\n\n${text}`
                    }
                ],
                max_tokens: 1000,
                temperature: 0.3
            });
            
            const summary = response.choices[0].message.content;
            console.log('✓ Summary generated successfully');
            return summary;
        } catch (error) {
            throw new Error(`Summary generation failed: ${error.message}`);
        }
    }

    /**
     * Extract frequently mentioned topics using GPT
     */
    async extractTopics(text) {
        console.log('Extracting topics...');
        
        try {
            const response = await this.client.chat.completions.create({
                model: 'gpt-4',
                messages: [
                    {
                        role: 'system',
                        content: `You are an expert at analyzing text and identifying key topics and themes.
                        Return a JSON array of the most frequently mentioned topics with their mention counts.
                        Each item should have 'topic' and 'mentions' fields.
                        Focus on meaningful topics, not common words like 'the', 'and', etc.
                        Return at least 3 topics, more if there are significant ones.`
                    },
                    {
                        role: 'user',
                        content: `Analyze this transcribed audio content and identify the most frequently mentioned topics with their approximate mention counts:\n\n${text}\n\nReturn only a JSON array in this format: ` + '[{"topic": "Topic Name", "mentions": 5}]'
                    }
                ],
                max_tokens: 500,
                temperature: 0.1
            });
            
            const topicsText = response.choices[0].message.content;
            // Extract JSON from the response
            const jsonMatch = topicsText.match(/\[.*\]/s);
            let topics;
            
            if (jsonMatch) {
                topics = JSON.parse(jsonMatch[0]);
            } else {
                // Fallback: try to parse the entire response as JSON
                topics = JSON.parse(topicsText);
            }
            
            console.log('✓ Topics extracted successfully');
            return topics;
        } catch (error) {
            console.warn(`Warning: Error extracting topics with GPT: ${error.message}`);
            // Fallback to simple topic extraction
            return this.simpleTopicExtraction(text);
        }
    }

    /**
     * Fallback simple topic extraction based on word frequency
     */
    simpleTopicExtraction(text) {
        console.log('Using fallback topic extraction...');
        
        // Clean and split text
        const words = text.toLowerCase().match(/\b[a-zA-Z]{3,}\b/g) || [];
        
        // Common stop words to filter out
        const stopWords = new Set([
            'the', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for', 'of', 'with',
            'by', 'from', 'as', 'is', 'was', 'are', 'were', 'be', 'been', 'have',
            'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could', 'should',
            'may', 'might', 'must', 'can', 'this', 'that', 'these', 'those', 'i',
            'you', 'he', 'she', 'it', 'we', 'they', 'me', 'him', 'her', 'us', 'them',
            'my', 'your', 'his', 'her', 'its', 'our', 'their', 'a', 'an', 'some',
            'any', 'all', 'each', 'every', 'no', 'not', 'very', 'so', 'just', 'now',
            'then', 'than', 'only', 'also', 'back', 'other', 'many', 'much', 'more',
            'most', 'good', 'new', 'first', 'last', 'long', 'great', 'little', 'own',
            'same', 'right', 'big', 'high', 'different', 'small', 'large', 'next',
            'early', 'young', 'important', 'few', 'public', 'bad', 'same', 'able'
        ]);
        
        // Count word frequencies
        const wordCounts = {};
        words.forEach(word => {
            if (!stopWords.has(word) && word.length > 3) {
                wordCounts[word] = (wordCounts[word] || 0) + 1;
            }
        });
        
        // Get top topics
        const sortedWords = Object.entries(wordCounts)
            .sort(([,a], [,b]) => b - a)
            .slice(0, 5);
        
        return sortedWords.map(([word, count]) => ({
            topic: word.charAt(0).toUpperCase() + word.slice(1),
            mentions: count
        }));
    }

    /**
     * Calculate comprehensive analytics for the transcribed text
     */
    async calculateAnalytics(text, audioDuration) {
        console.log('Calculating analytics...');
        
        // Word count
        const words = text.split(/\s+/).filter(word => word.length > 0);
        const wordCount = words.length;
        
        // Speaking speed (WPM)
        let speakingSpeed = 0;
        if (audioDuration > 0) {
            const durationMinutes = audioDuration / 60;
            speakingSpeed = Math.round((wordCount / durationMinutes) * 10) / 10;
        }
        
        // Extract topics
        const topics = await this.extractTopics(text);
        
        const analytics = {
            word_count: wordCount,
            speaking_speed_wpm: speakingSpeed,
            audio_duration_seconds: Math.round(audioDuration * 100) / 100,
            frequently_mentioned_topics: topics
        };
        
        console.log('✓ Analytics calculated successfully');
        return analytics;
    }

    /**
     * Save transcription, summary, and analytics to separate files
     */
    async saveFiles(transcription, summary, analytics, originalFilename) {
        const baseName = path.parse(originalFilename).name;
        const timestamp = new Date().toISOString().replace(/[:.]/g, '-').split('T')[0] + '_' + 
                         new Date().toTimeString().split(' ')[0].replace(/:/g, '');
        
        // Create a folder for this audio file
        let audioFolderName = baseName;
        let audioFolderPath = path.join(this.outputDir, audioFolderName);
        
        // Handle duplicate folder names by adding a number
        let counter = 1;
        while (true) {
            try {
                await fs.access(audioFolderPath);
                // Folder exists, try with counter
                audioFolderName = `${baseName}_${counter}`;
                audioFolderPath = path.join(this.outputDir, audioFolderName);
                counter++;
            } catch (error) {
                // Folder doesn't exist, we can use this name
                break;
            }
        }
        
        // Create the audio-specific folder
        await fs.mkdir(audioFolderPath, { recursive: true });
        
        // File paths (simple names within the folder)
        const transcriptionPath = path.join(audioFolderPath, 'transcription.md');
        const summaryPath = path.join(audioFolderPath, 'summary.md');
        const analyticsPath = path.join(audioFolderPath, 'analysis.json');
        
        // Save transcription
        const transcriptionContent = `# Transcription\n\n` +
            `**File:** ${originalFilename}\n` +
            `**Date:** ${new Date().toLocaleString()}\n` +
            `**Processing ID:** ${timestamp}\n\n` +
            `## Content\n\n${transcription}\n`;
        await fs.writeFile(transcriptionPath, transcriptionContent, 'utf8');
        
        // Save summary
        const summaryContent = `# Summary\n\n` +
            `**File:** ${originalFilename}\n` +
            `**Date:** ${new Date().toLocaleString()}\n` +
            `**Processing ID:** ${timestamp}\n\n` +
            `## Summary\n\n${summary}\n`;
        await fs.writeFile(summaryPath, summaryContent, 'utf8');
        
        // Add metadata to analytics
        const analyticsWithMeta = {
            ...analytics,
            metadata: {
                original_filename: originalFilename,
                processed_date: new Date().toISOString(),
                processing_id: timestamp,
                output_folder: audioFolderName
            }
        };
        await fs.writeFile(analyticsPath, JSON.stringify(analyticsWithMeta, null, 2), 'utf8');
        
        console.log('✓ Files saved:');
        console.log(`  - Folder: ${audioFolderPath}`);
        console.log(`  - Transcription: ${transcriptionPath}`);
        console.log(`  - Summary: ${summaryPath}`);
        console.log(`  - Analytics: ${analyticsPath}`);
        
        return { 
            folderPath: audioFolderPath,
            transcriptionPath, 
            summaryPath, 
            analyticsPath 
        };
    }

    /**
     * Main method to process audio file completely
     */
    async processAudio(audioPath) {
        console.log('\n' + '='.repeat(60));
        console.log('AUDIO TRANSCRIPTION AND ANALYSIS');
        console.log('='.repeat(60));
        console.log(`Processing: ${audioPath}`);
        console.log('='.repeat(60) + '\n');
        
        try {
            // Validate audio file (format, size, existence)
            await this.validateAudioFile(audioPath);
            
            // Get audio duration
            const audioDuration = await this.getAudioDuration(audioPath);
            console.log(`Audio duration: ${audioDuration.toFixed(2)} seconds\n`);
            
            // Step 1: Transcribe audio
            const transcription = await this.transcribeAudio(audioPath);
            
            // Step 2: Generate summary
            const summary = await this.summarizeText(transcription);
            
            // Step 3: Calculate analytics
            const analytics = await this.calculateAnalytics(transcription, audioDuration);
            
            // Step 4: Save files
            await this.saveFiles(transcription, summary, analytics, path.basename(audioPath));
            
            // Step 5: Display results
            this.displayResults(transcription, summary, analytics);
            
        } catch (error) {
            console.error(`❌ Error processing audio: ${error.message}`);
            process.exit(1);
        }
    }

    /**
     * Display results in the console
     */
    displayResults(transcription, summary, analytics) {
        console.log('\n' + '='.repeat(60));
        console.log('RESULTS');
        console.log('='.repeat(60));
        
        console.log('\n📊 ANALYTICS:');
        console.log('='.repeat(30));
        console.log(JSON.stringify(analytics, null, 2));
        
        console.log('\n📝 SUMMARY:');
        console.log('='.repeat(30));
        console.log(summary);
        
        console.log('\n📄 TRANSCRIPTION (first 500 chars):');
        console.log('='.repeat(30));
        const truncatedTranscription = transcription.length > 500 
            ? transcription.substring(0, 500) + '...'
            : transcription;
        console.log(truncatedTranscription);
        
        console.log('\n' + '='.repeat(60));
        console.log('✅ Processing completed successfully!');
        console.log('='.repeat(60));
    }
}

/**
 * Main function to handle command line arguments and run the application
 */
async function main() {
    program
        .name('audio-analyzer')
        .description('Audio Transcription and Analysis Tool')
        .version('1.0.0')
        .argument('<audio_file>', 'Path to the audio file to transcribe and analyze')
        .option('--api-key <key>', 'OpenAI API key (can also be set via OPENAI_API_KEY environment variable)')
        .action(async (audioFile, options) => {
            // Get API key
            const apiKey = options.apiKey || process.env.OPENAI_API_KEY;
            if (!apiKey) {
                console.error('❌ Error: OpenAI API key required!');
                console.error('Set it via:');
                console.error('  - Environment variable: export OPENAI_API_KEY="your-key-here"');
                console.error('  - Command line argument: --api-key YOUR_KEY');
                process.exit(1);
            }
            
            // Validate audio file exists
            try {
                await fs.access(audioFile);
            } catch (error) {
                console.error(`❌ Error: Audio file not found: ${audioFile}`);
                process.exit(1);
            }
            
            // Process audio
            const analyzer = new AudioAnalyzer(apiKey);
            await analyzer.processAudio(audioFile);
        });

    program.parse();
}

// Handle unhandled promise rejections
process.on('unhandledRejection', (error) => {
    console.error('❌ Unhandled error:', error.message);
    process.exit(1);
});

// Run the application
if (require.main === module) {
    main().catch(error => {
        console.error('❌ Application error:', error.message);
        process.exit(1);
    });
}

module.exports = { AudioAnalyzer };